crates/ruff_server/src/server/api/requests/code_action.rs

```diff
@@ -1,81 +1,129 @@
+            todo!("Implement the `source.organizeImports` code action");
+        }
+
+        Ok(Some(response))
+    }
+}
+
+fn quick_fix(
+    snapshot: &DocumentSnapshot,
+    diagnostics: Vec<types::Diagnostic>,
+) -> crate::Result<impl Iterator<Item = CodeActionOrCommand> + '_> {
+    let document = snapshot.document();
+
+    let fixes = fixes_for_diagnostics(
+        document,
+        snapshot.url(),
+        snapshot.encoding(),
+        document.version(),
+        diagnostics,
+    )?;
+
+    Ok(fixes.into_iter().map(|fix| {
+        types::CodeActionOrCommand::CodeAction(types::CodeAction {
+            title: format!("{DIAGNOSTIC_NAME} ({}): {}", fix.code, fix.title),
+            kind: Some(types::CodeActionKind::QUICKFIX),
+            edit: Some(types::WorkspaceEdit {
+                document_changes: Some(types::DocumentChanges::Edits(fix.document_edits.clone())),
+                ..Default::default()
+            }),
+            diagnostics: Some(vec![fix.fixed_diagnostic.clone()]),
+            data: Some(serde_json::to_value(snapshot.url()).expect("document url to serialize")),
+            ..Default::default()
+        })
+    }))
+}
+
+fn fix_all(snapshot: &DocumentSnapshot) -> crate::Result<CodeActionOrCommand> {
+    let mut action = types::CodeAction {
+        title: format!("{DIAGNOSTIC_NAME}: Fix all auto-fixable problems"),
+        kind: Some(types::CodeActionKind::SOURCE_FIX_ALL),
+        // This will be resolved later
+        edit: None,
+        data: Some(serde_json::to_value(snapshot.url()).expect("document url to serialize")),
+        ..Default::default()
+    };
+
+    if !snapshot
+        .resolved_client_capabilities()
+        .code_action_deferred_edit_resolution
+    {
         let document = snapshot.document();
-        let url = snapshot.url();
-        let encoding = snapshot.encoding();
-        let version = document.version();
-        let actions: Result<Vec<_>> = params
-            .context
-            .diagnostics
-            .into_iter()
-            .map(|diagnostic| {
-                let Some(data) = diagnostic.data else {
-                    return Ok(None);
-                };
-                let diagnostic_fix: crate::lint::DiagnosticFix = serde_json::from_value(data)
-                    .map_err(|err| anyhow::anyhow!("failed to deserialize diagnostic data: {err}"))
-                    .with_failure_code(lsp_server::ErrorCode::ParseError)?;
-                let edits = diagnostic_fix
-                    .fix
-                    .edits()
+
+        // We need to resolve the `edit` field now if we can't defer resolution to later
+        action = resolve_edit_for_fix_all(
+            action,
+            document,
+            snapshot.url(),
+            &snapshot.configuration().linter,
+            snapshot.encoding(),
+        )?;
+    }
+
+    Ok(types::CodeActionOrCommand::CodeAction(action))
```

Comment:
It's a bit confusing that we initialize `action` only to then override it. To me it seems the only motivation for doing this is so that the `action` can be passed to `resolve_edit_for_fix_all`, which sets `action.edit`. I think I would restructure this to
```suggestion
    let (edit, data) = if snapshot.resolved_client_capabilities().code_action_deferred_edit_resolution {
        // This will be resolved later
        (None, Some(serde_json::to_value(snapshot.url()).expect("document url to serialize")))
    } else {
        (resolve_edit_for_fix_all(document, snapshot.url(), &snapshot.configuration.linter, snapshot.encoding()), None)
    };

    let action = types::CodeAction {
        title: format!("{DIAGNOSTIC_NAME}: Fix all auto-fixable problems"),
        kind: Some(types::CodeActionKind::SOURCE_FIX_ALL),
        edit: None,
        data,
        ..Default::default()
    };

    Ok(types::CodeActionOrCommand::CodeAction(action))
```

This way it's clearer what's different between the two paths and it avoids setting `data` for clients that don't support it.